lib/Auth/Process/privacyidea.php

```diff
@@ -104,67 +104,77 @@ public function process(&$state)
-	        $status = $result->status;
-	        $value = $result->value->auth;
-        } catch (Exception $e) {
-	        throw new SimpleSAML_Error_BadRequest( "privacyIDEA: We were not able to read the response from the PI server");
+        if ($transaction_id) {
+            SimpleSAML_Logger::debug("Authenticating with transaction_id: " . $transaction_id);
+            $params["transaction_id"] = $transaction_id;
         }
-
-        if ($status !== True) {
-            throw new SimpleSAML_Error_BadRequest("Valid JSON Response, but some internal error occured in PI server");
-        } else {
-            if ($value !== True){
-                SimpleSAML_Logger::debug("privacyIDEA: Wrong user pass");
-                return False;
-            } else {
-	            SimpleSAML_Logger::debug("privacyIDEA: User authenticated successfully");
-                return True;
-            }
+        if ($signaturedata) {
+            SimpleSAML_Logger::debug("Authenticating with signaturedata: " . urlencode($signaturedata));
+            $params["signaturedata"] = $signaturedata;
+        }
+        if ($clientdata) {
+            SimpleSAML_Logger::debug("Authenticating with clientdata: " . urlencode($clientdata));
+            $params["clientdata"] = $clientdata;
         }
+	    $body = sspmod_privacyidea_Auth_utils::curl($params, null, $cfg, "/validate/samlcheck", true);
+		$attributes = NULL;
 
+	    try {
+		    $result = $body->result;
+		    $status = $result->status;
+		    $value  = $result->value;
+		    $auth   = $value->auth;
+	    } catch (Exception $e) {
+		    throw new SimpleSAML_Error_BadRequest("privacyIDEA: We were not able to read the response from the PI server");
+	    }
 
+	    if ($status !== true) {
+		    throw new SimpleSAML_Error_BadRequest("privacyIDEA: Valid JSON response, but some internal error occured in PI server");
+	    } else {
+		    if ( $auth !== true ) {
+			    SimpleSAML_Logger::debug( "Throwing WRONGUSERPASS" );
+			    $detail = $body->detail;
+			    $message = $detail->message;
+			    if (property_exists($detail, "transaction_id")){
+				    $transaction_id = $detail->transaction_id;
+			    }
+			    if (property_exists( $detail, "attributes")){
+				    $attributes = $detail->attributes;
+				    if (property_exists( $attributes, "u2fSignRequest")){
+					    SimpleSAML_Logger::debug("This is an U2F authentication request");
+					    SimpleSAML_Logger::debug(print_r($attributes, true));
+					    /*
+						 * In case of U2F the $attributes looks like this:
+						[img] => static/css/FIDO-U2F-Security-Key-444x444.png#012
+						[hideResponseInput] => 1#012
+						[u2fSignRequest] => [challenge] => yji-PL1V0QELilDL3m6Lc-1yahpKZiU-z6ye5Zz2mp8#012
+									[version] => U2F_V2#012
+									[keyHandle] => fxDKTr6o8EEGWPyEyRVDvnoeA0c6v-dgvbN-6Mxc6XBmEItsw#012
+									[appId] => https://172.16.200.138#012        )#012#012)
+						*/
+				    }
+			    }
+			    if ($transaction_id){
```

Comment:
should we add the lines 148 - 150 in this if-branch, so that we only have two if-branches on this level, and not three.
(we check for the transaction_id two times)

Reply:
fixed in 8f80d94